Sources/SwiftKafka/KafkaConsumer.swift

```diff
@@ -0,0 +1,418 @@
+    /// - Parameter topic: Name of the topic that this ``KafkaConsumer`` will read from.
+    /// - Parameter partition: Partition that this ``KafkaConsumer`` will read from.
+    /// - Parameter config: The ``KafkaConfig`` for configuring the ``KafkaConsumer``.
+    /// - Parameter logger: A logger.
+    /// - Note: This consumer ignores the `group.id` property of its `config`.
+    public convenience init(
+        topic: String,
+        partition: KafkaPartition,
+        config: KafkaConfig = KafkaConfig(),
+        logger: Logger
+    ) async throws {
+        // Althogh an assignment is not related to a consumer group,
+        // librdkafka requires us to set a `group.id`.
+        // This is a known issue:
+        // https://github.com/edenhill/librdkafka/issues/3261
+        var config = config
+        try config.set(UUID().uuidString, forKey: "group.id")
+
+        try await self.init(
+            config: config,
+            logger: logger
+        )
+        try await self.assign(topic: topic, partition: partition)
+    }
+
+    deinit {
+        switch self.state {
+        case .started:
+            Task {
+                try await self._close()
+            }
+        case .closed:
+            return
+        }
+    }
+
+    /// Subscribe to the given list of `topics`.
+    /// The partition assignment happens automatically using `KafkaConsumer`'s consumer group.
+    /// - Parameter topics: An array of topic names to subscribe to.
+    private func subscribe(topics: [String]) throws {
+        // TODO: is this state needed for a method that is only invoked upon init?
+        switch self.state {
+        case .started:
+            break
+        case .closed:
+            throw KafkaError(description: "Trying to invoke method on consumer that has been closed.")
+        }
+
+        for topic in topics {
+            rd_kafka_topic_partition_list_add(
+                self.subscribedTopicsPointer,
+                topic,
+                KafkaPartition.unassigned.rawValue
+            )
+        }
+
+        let result = self.client.withKafkaHandlePointer { handle in
+            rd_kafka_subscribe(handle, subscribedTopicsPointer)
+        }
+
+        guard result.rawValue == 0 else {
+            throw KafkaError(rawValue: result.rawValue)
+        }
+    }
+
+    /// Assign the``KafkaConsumer`` to a specific `partition` of a `topic`.
+    /// - Parameter topic: Name of the topic that this ``KafkaConsumer`` will read from.
+    /// - Parameter partition: Partition that this ``KafkaConsumer`` will read from.
+    /// - Parameter offset: The topic offset where reading begins. Defaults to the offset of the last read message.
+    private func assign(
+        topic: String,
+        partition: KafkaPartition,
+        offset: Int64 = Int64(RD_KAFKA_OFFSET_END)
+    ) throws {
+        // TODO: is this state needed for a method that is only invoked upon init?
```

Comment:
Currently, `assign` is only invoked by `init`. Thus we could potentially get rid of the state. But this totally depends on if we want to allow users to `subscribe/assign` to more topics during runtime.